File: src/familiar_connect/tools/waker.py
"""Alarm waker processor.

Listens on :data:`TOPIC_ALARM_FIRED`, republishes a synthetic
``discord.text``-shaped event so existing :class:`TextResponder`
picks it up and produces a follow-up reply.

MVP behavior:

* text-origin alarms → publish synthetic ``discord.text`` event with
  ``content = "[alarm fired: {reason}]"``, ``author=None``.
* voice-origin alarms → fall back to text by publishing same shape
  using the alarm's recorded ``channel_id``. Real Discord voice and
  text channels have distinct ids — production wiring needs a
  per-channel mapping (out of scope for MVP).
"""

from __future__ import annotations

import logging
import uuid
from datetime import UTC, datetime
from typing import TYPE_CHECKING

from familiar_connect import log_style as ls
from familiar_connect.bus.envelope import Event
from familiar_connect.bus.topics import TOPIC_ALARM_FIRED, TOPIC_DISCORD_TEXT

if TYPE_CHECKING:
    from familiar_connect.bus.protocols import EventBus

_logger = logging.getLogger(__name__)


class AlarmWaker:
    """Translate ``alarm.fired`` into a synthetic text-channel turn."""

    name: str = "alarm-waker"
    topics: tuple[str, ...] = (TOPIC_ALARM_FIRED,)

    def __init__(self, *, familiar_id: str) -> None:
        self._familiar_id = familiar_id

    async def handle(self, event: Event, bus: EventBus) -> None:
        if event.topic != TOPIC_ALARM_FIRED:
            return
        payload = event.payload
        if not isinstance(payload, dict):
            return

        channel_id = payload.get("channel_id")
        reason = payload.get("reason") or ""
        if not isinstance(channel_id, int):
            return

        kind = payload.get("channel_kind") or "text"
        if kind not in {"text", "voice"}:
            _logger.warning(
                f"{ls.tag('AlarmWaker', ls.LY)} "
                f"{ls.kv('unknown_channel_kind', str(kind), vc=ls.LY)}"
            )
            return

        if kind == "voice":
            _logger.info(
                f"{ls.tag('AlarmWaker', ls.LM)} "
                f"{ls.kv('voice_fallback_to_text', str(channel_id), vc=ls.LM)} "
                f"{ls.kv('reason', ls.trunc(reason, limit=80), vc=ls.LW)}"
            )

        synth_event_id = uuid.uuid4().hex
        synth_payload = {
            "familiar_id": self._familiar_id,
            "channel_id": channel_id,
            "content": f"[alarm fired: {reason}]",
            "author": None,
            "guild_id": None,
            "message_id": None,
            "reply_to_message_id": None,
            "mentions": (),
        }
        await bus.publish(
            Event(
                event_id=synth_event_id,
                turn_id=f"wake-{payload.get('alarm_id', synth_event_id)}",
                session_id=str(channel_id),
                parent_event_ids=(event.event_id,),
                topic=TOPIC_DISCORD_TEXT,
                timestamp=datetime.now(tz=UTC),
                sequence_number=0,
                payload=synth_payload,
            )
        )

File: rust/familiar-connect/src/tools/waker.rs
//! Alarm waker processor (subsystem 08; Python `tools/waker.py`).
//!
//! Listens on [`TOPIC_ALARM_FIRED`] and republishes a synthetic
//! `discord.text`-shaped event so the existing `TextResponder` picks it up and
//! produces a follow-up reply. Voice-origin alarms fall back to text in the same
//! channel id (MVP). The synthetic payload carries the *waker's* configured
//! `familiar_id` (alarm payloads carry none); per-familiar filtering happens in
//! the responder. The `alarm: true` marker pierces activity absence gating.

use async_trait::async_trait;
use serde_json::{Value, json};

use crate::bus::envelope::{Event, payload};
use crate::bus::protocols::{EventBus, Processor};
use crate::bus::topics::{TOPIC_ALARM_FIRED, TOPIC_DISCORD_TEXT};
use crate::log_style as ls;

/// Translate `alarm.fired` into a synthetic text-channel turn.
pub struct AlarmWaker {
    familiar_id: String,
    topics: [&'static str; 1],
}

impl AlarmWaker {
    /// New waker bound to a familiar id.
    #[must_use]
    pub fn new(familiar_id: impl Into<String>) -> Self {
        Self {
            familiar_id: familiar_id.into(),
            topics: [TOPIC_ALARM_FIRED],
        }
    }
}

#[async_trait]
impl Processor for AlarmWaker {
    #[allow(
        clippy::unnecessary_literal_bound,
        reason = "return type is fixed by the Processor trait signature"
    )]
    fn name(&self) -> &str {
        "alarm-waker"
    }

    fn topics(&self) -> &[&str] {
        &self.topics
    }

    async fn handle(&self, event: std::sync::Arc<Event>, bus: &dyn EventBus) -> anyhow::Result<()> {
        if event.topic != TOPIC_ALARM_FIRED {
            return Ok(());
        }
        let Some(payload_val) = event.payload.downcast_ref::<Value>() else {
            return Ok(());
        };
        if !payload_val.is_object() {
            return Ok(());
        }

        let Some(channel_id) = payload_val.get("channel_id").and_then(Value::as_i64) else {
            return Ok(());
        };
        let reason = payload_val
            .get("reason")
            .and_then(Value::as_str)
            .unwrap_or("");

        let kind = payload_val
            .get("channel_kind")
            .and_then(Value::as_str)
            .filter(|s| !s.is_empty())
            .unwrap_or("text");
        if kind != "text" && kind != "voice" {
            tracing::warn!(
                "{} {}",
                ls::tag("AlarmWaker", ls::LY),
                ls::kv_styled("unknown_channel_kind", kind, ls::W, ls::LY),
            );
            return Ok(());
        }
        if kind == "voice" {
            tracing::info!(
                "{} {} {}",
                ls::tag("AlarmWaker", ls::LM),
                ls::kv_styled(
                    "voice_fallback_to_text",
                    &channel_id.to_string(),
                    ls::W,
                    ls::LM
                ),
                ls::kv_styled("reason", &ls::trunc(reason, 80), ls::W, ls::LW),
            );
        }

        let synth_event_id = uuid::Uuid::new_v4().simple().to_string();
        let alarm_id = payload_val
            .get("alarm_id")
            .and_then(Value::as_str)
            .unwrap_or(&synth_event_id);
        let synth_payload: Value = json!({
            "familiar_id": self.familiar_id,
            "channel_id": channel_id,
            "content": format!("[alarm fired: {reason}]"),
            "author": null,
            "guild_id": null,
            "message_id": null,
            "reply_to_message_id": null,
            "mentions": [],
            "alarm": true,
        });

        bus.publish(Event {
            event_id: synth_event_id.clone(),
            turn_id: format!("wake-{alarm_id}"),
            session_id: channel_id.to_string(),
            parent_event_ids: vec![event.event_id.clone()],
            topic: TOPIC_DISCORD_TEXT.to_owned(),
            timestamp: chrono::Utc::now(),
            sequence_number: 0,
            payload: payload(synth_payload),
        })
        .await;
        Ok(())
    }
}
